lms/lmsdb/models.py

```diff
@@ -431,25 +482,27 @@ def test_results(self) -> Iterable[dict]:
     @classmethod
     def of_user(
         cls, user_id: int, with_archived: bool = False,
-    ) -> Iterable[Dict[str, Any]]:
-        db_exercises = Exercise.get_objects(fetch_archived=with_archived)
+    ) -> Iterable[DefaultDict[str, Dict[str, Any]]]:
+        db_exercises = Exercise.get_objects(
+            user_id=user_id, fetch_archived=with_archived,
+        )
         exercises = Exercise.as_dicts(db_exercises)
-
         solutions = (
             cls
             .select(cls.exercise, cls.id, cls.state, cls.checker)
             .where(cls.exercise.in_(db_exercises), cls.solver == user_id)
             .order_by(cls.submission_timestamp.desc())
         )
         for solution in solutions:
-            exercise = exercises[solution.exercise_id]
+            course_name = solution.exercise.course.name
+            exercise = exercises[course_name][solution.exercise_id]
```

Comment:
Be sure to cover this one thoroughly with tests. Nested structures tend to have hard to detect bugs.

Reply:
Changed it back